database/triggers.py

```diff
@@ -0,0 +1,213 @@
+from alembic_utils.pg_function import PGFunction
+from alembic_utils.pg_trigger import PGTrigger
+
+hame_tables = [
+    "plan",
+    "land_use_area",
+    "other_area",
+    "line",
+    "land_use_point",
+    "other_point",
+    "plan_regulation_group",
+    "plan_regulation",
+    "plan_proposition",
+    "source_data",
+    "organisation",
+    "document",
+    "lifecycle_date",
+]
+
+tables_with_lifecycle_date = [
+    "plan",
+    "plan_regulation",
+    "plan_proposition",
+]
+
+tables_with_lifecycle_status = [
+    # plan and lifecycle_date do not belong here
+    "land_use_area",
+    "other_area",
+    "line",
+    "land_use_point",
+    "other_point",
+    "plan_regulation",
+    "plan_proposition",
+]
+
+plan_object_tables = [
+    "land_use_area",
+    "other_area",
+    "line",
+    "land_use_point",
+    "other_point",
+]
+
+
+def generate_modified_at_triggers():
+    modified_at_trgs = []
+    modified_at_trgfuncs = []
+
+    for table in hame_tables:
+        trgfunc_signature = f"trgfunc_{table}_modified_at()"
+        trgfunc_definition = """
+        RETURNS TRIGGER AS $$
+        BEGIN
+            NEW.modified_at = CURRENT_TIMESTAMP;
+            return NEW;
+        END;
+        $$ language 'plpgsql'
+        """
+
+        trg_signature = f"trg_{table}_modified_at"
+        trg_definition = f"""
+        BEFORE INSERT OR UPDATE ON {table}
+        FOR EACH ROW
+        EXECUTE FUNCTION hame.{trgfunc_signature}
+        """
+
+        trgfunc = PGFunction(
+            schema="hame",
+            signature=trgfunc_signature,
+            definition=trgfunc_definition,
+        )
+        modified_at_trgfuncs.append(trgfunc)
+
+        trg = PGTrigger(
+            schema="hame",
+            signature=trg_signature,
+            on_entity=f"hame.{table}",
+            is_constraint=False,
+            definition=trg_definition,
+        )
+        modified_at_trgs.append(trg)
+
+    return modified_at_trgs, modified_at_trgfuncs
+
+
+def generate_new_lifecycle_date_triggers():
+    new_lifecycle_date_trgs = []
+    new_lifecycle_date_trgfuncs = []
+
+    for table in tables_with_lifecycle_date:
+        trgfunc_signature = f"trgfunc_{table}_new_lifecycle_date()"
+        trgfunc_definition = f"""
+        RETURNS TRIGGER AS $$
+        BEGIN
+            INSERT INTO hame.lifecycle_date
+                (lifecycle_status_id, {table}_id, starting_at)
+            VALUES (NEW.lifecycle_status_id, NEW.id, CURRENT_TIMESTAMP);
+            RETURN NEW;
+        END;
```

Comment:
@Rikuoja What's the reason for setting `ending_at` to have the current timestamp? Is the end date of a lifecycle status not something that the planner sets to some predefined point in the future?

Reply:
@msorvoja No, the status ends exactly when it is changed! A plan object (or plan, or regulation) can only be in one state at any given time. When the status changes (e.g. from Kaavaehdotus to Hyväksytty), the old status is no longer valid.

In some cases, it might be possible that the planner may set the end date in Voimassa to a specific date, if we know that the status changes again at that point. But usually we don't know beforehand when the status will change.

An exception might be from Hyväksytty to Voimassa. In that case, the planner might first change status Hyväksytty -> Voimassa, and then change the generated dates so that Voimassa only starts later.

But in general, let's first try out this way of working. And then if in some cases we want to set the start and end dates manually, we can disable the trigger for some status changes.